worlds/animal_well/patch.py

```diff
@@ -0,0 +1,179 @@
+
+        self.original_bytes = self.process.read_bytes(self.base_address, len(self.byte_list))
+        self.process.write_bytes(self.base_address, self.byte_list, len(self.byte_list))
+
+        if self.process.read_bytes(self.base_address, len(self.byte_list)) != self.byte_list:
+            print(f'Failed to apply patch {self.name} ({len(self)} length) at {hex(self.base_address)}!')
+            return False
+
+        print(f'Patch {self.name} ({len(self)} length) applied at {hex(self.base_address)} successfully!')
+        self.patch_applied = True
+        return True
+
+    def revert(self):
+        if self.process == None:
+            print(f'Failed to apply patch {self.name}, process has not been specified!')
+            return False
+
+        if not self.patch_applied:
+            print(f'Failed to revert patch {self.name}, patch has not yet been applied!')
+            return False
+
+        if self.original_bytes == None or len(self.original_bytes) == 0:
+            print(f'OriginalBytes array is blank. Cannot revert patch {self.name}!')
+            return False
+
+        self.process.write_bytes(self.base_address, self.original_bytes, len(self.original_bytes))
+
+        if self.process.read_bytes(self.base_address, len(self.original_bytes)) != self.original_bytes:
+            print(f'Failed to revert patch {self.name} ({len(self)} length) at {hex(self.base_address)}!')
+            return False
+
+        print(f'Patch {self.name} ({len(self)} length) at {hex(self.base_address)} reverted successfully!')
+        self.patch_applied = False
+        self.original_bytes = b''
+        return True
+
+    def add_bytes(self, bytes):
+        self.byte_list += bytes
+        return self
+
+    def xor_ecx_ecx(self): #2 bytes
+        self.add_bytes(b'\x31\xc9')
+        return self
+
+    def call_near(self, address): #5 bytes
+        diff = address - (self.get_patch_end() + 5)
+        print('diff: {} - {} = {}'.format(hex(address), hex(self.get_patch_end() + 5), hex(diff)))
+        self.add_bytes(b'\xe8' + (diff).to_bytes(4, 'little', signed=True))
+        return self
+
+    def call_far(self, address): #16 bytes
+        self.add_bytes(b'\xff\x15\x02\x00\x00\x00\xeb\x08' + address.to_bytes(8, 'little', signed=True))
+        return self
+
+    def call_rax(self): #2
+        self.add_bytes(b'\xff\xd0')
+        return self
+
+    def jmp_near(self, address): #5 bytes
+        diff = address - (self.get_patch_end() + 5)
+        self.add_bytes(b'\xe9' + (diff).to_bytes(4, 'little', signed=True))
+        return self
+
+    def jmp_far(self, address): #14 bytes
+        self.add_bytes(b'\xff\x25\x00\x00\x00\x00' + address.to_bytes(8, 'little', signed=True))
+        return self
+
+    def nop(self, count):
+        self.add_bytes(b'\x90' * count)
+        return self
+
+    def mov_ecx(self, value): #5
+        self.add_bytes(b'\xb9' + value.to_bytes(4, 'little'))
+        return self
+
+    def mov_edx(self, value): #5
```

Comment:
Definitely need some sort of comments on these -- I have no idea what these are doing

Reply:
Yeah, I can try and add some documentation to the Patch class. The overwhelming majority of those functions are assembly instructions, so the documentation will help a little but a lot of understanding what those do is going to rely on your familiarity with x86 assembly.

And I was hoping that you'd be willing to improve the formatting on stuff. This being my first work on an AP Client (or anything AP related), and me primarily being a C# dev professionally, Python and this entire project are pretty new territory for me, so some things about how I should format everything and keep it `Pythonic` are gonna take time for me to learn. I'm thinking that a lot of the new information I'm storing directly in the context would probably do well to be wrapped up into categories of things. There are also a lot of constants involved in this, since I'm currently using hard-coded addresses instead of AOB patterns to find functions, locations to patch, etc, so any way that we can group those together and limit how much they take up space inside of the client's main code, I'm all for.

Reply:
TONS of documentation added for all the Patch stuff. Also thoroughly documented the patches themselves. Those patches have been moved to Bean_Patcher, which now contains all of the ASM patches themselves, and keep all of the state information that I was previously keeping directly in the context.